query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::rc::Rc;
+
+use mentat_core::{
+    Schema,
+    TypedValue,
+    ValueType,
+};
+
+use mentat_query::{
+    Binding,
+    FnArg,
+    NonIntegerConstant,
+    SrcVar,
+    Variable,
+    WhereFn,
+};
+
+use clauses::{
+    ConjoiningClauses,
+    PushComputed,
+};
+
+use errors::{
+    BindingError,
+    ErrorKind,
+    Result,
+};
+
+use types::{
+    ComputedTable,
+    SourceAlias,
+    VariableColumn,
+};
+
+trait FnArgInto<F> {
+    fn into_typed_value(self, lookup: F) -> Result<TypedValue>;
+    fn into_column_vector(self, lookup: F) -> Result<Vec<TypedValue>>;
+
+    fn into_unit_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+    fn into_row_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+    fn into_column_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+    fn into_block_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+}
+
+impl<F> FnArgInto<F> for FnArg where F: Fn(Variable) -> Result<TypedValue> {
+    fn into_typed_value(self, lookup: F) -> Result<TypedValue> {
+        match self {
+            FnArg::EntidOrInteger(x) => Ok(TypedValue::Long(x)),
```

Comment:
Here's my proposed approach.

Datomic discusses entity resolution in the 'v' place:

http://docs.datomic.com/query.html#sec-8-1

and this problem is similar. If, for ambiguous values, we can delay concretization until all other patterns have been algebrized, then we can make a type-aware determination, falling back to a default if necessary.

If a user wants an ident to always resolve to an entid, never to a keyword, then they can do so themselves. Our algorithm, then:

1. If a value is of an unambiguous type (e.g., instant), then we're done.
2. If the value is of an ambiguous type (e.g., ident) but can't resolve as one kind, then it must be the other — `:foo/notattribute` is a keyword. We have the schema at alg time.
3. If the value is of an ambiguous type, but the query determines the type (e.g., by use of a keyword in the entity or attribute place), then we're done.
4. Finally, default to a single type.

We might consider supporting an `#entid` EDN feature.